simexpal/base.py

```diff
@@ -234,144 +183,79 @@ def get_build(self, name, revision):
-			# Determine all experiments selected by this scope.
-			if scope.selected_exps is None:
-				experiments = [exp_yml['name'] for exp_yml in self.yml['experiments']]
-			else:
-				experiments = scope.selected_exps
-
+		def extract(selection):
 			# Helper to find all selected revisions for a given experiment.
-			def revisions_for_experiment(exp_name):
-				if scope.selected_revs is None:
-					if 'use_builds' in self.get_experiment_info(exp_name)._exp_yml:
-						for revision in self.all_revisions():
-							yield revision.name
+			def revisions_for_experiment(exp_info):
+				if 'use_builds' in exp_info._exp_yml:
+					if [rev.name for rev in selection.revisions] != ['_none']:
+						yield from selection.revisions
 					else:
-						yield None
+						yield from self.all_revisions()
 				else:
-					yield from scope.selected_revs
+					yield Revision(self, {'name': '_none'})
 
-			variation_ids = self._expand_variation_ids_in_scope(scope)
-			for exp_name in experiments:
-				yield from itertools.product([exp_name], revisions_for_experiment(exp_name),
-						variation_ids)
+			for exp_info in selection.experiments:
+				yield from itertools.product([exp_info], revisions_for_experiment(exp_info),
+						selection.variations)
 
-		expansion = MatrixScope.walk_matrix(self, self.yml, expand)
-		return sorted(expansion)
-
-	def _experiment_matches_item(self, item_yml, name, revision):
-		if 'experiments' in item_yml:
-			if name not in item_yml['experiments']:
-				return False
-		if revision is None:
-			if 'revisions' in item_yml:
-				return False
-		else:
-			if 'revisions' in item_yml:
-				if revision.name not in item_yml['revisions']:
-					return False
-		return True
+		key = lambda x: (x[0].name, x[1].name, [sub_var.name for sub_var in x[2]])
+		for experiment_info, revision, variation in self._expand_matrix(extract, key=key):
+			revision = revision if revision.name != '_none' else None
+			yield Experiment(self, experiment_info, revision, variation)
 
 	def discover_all_runs(self):
 		for exp in self.all_experiments():
-			for inst_name, rep in self._expand_run_matrix(exp):
-				instance = self.get_instance(inst_name)
-				yield Run(self, exp, instance, rep)
-
-	def _expand_run_matrix(self, exp):
-		def expand(scope):
-			if scope.selected_exps is not None:
-				if exp.name not in scope.selected_exps:
-					return
-			if scope.selected_revs is not None:
-				if exp.revision.name not in scope.selected_revs:
-					return
-			if scope.selected_variants is not None:
-				variation_id = tuple(var.name for var in exp.variation)
-				if variation_id not in self._expand_variation_ids_in_scope(scope):
+			def extract(selection):
+				# TODO (for Duy): in a new commit, refactor this.
+				#      We expand the matrix once for each experiment. This is inefficient.
+				#      Instead, let this extract() function return tuples of
+				#      (experiment, revision, variation, instance, repetition)
+				#      and do only one _expand_matrix().
```

Comment:
What about this TODO? The remainder of the PR looks good to me.

Reply:
I added a new commit for this